typed-racket-lib/typed-racket/rep/type-rep.rkt

```diff
@@ -759,63 +788,156 @@
-   (intern-single-ref! intersection-table
-                       elems
-                       #:construct (make-Intersection ts elems))])
+  [#:custom-constructor/contract
+   (-> (cons/c Type? (listof Type?))
+       Prop?
+       (hash/c Type? #t #:immutable #t #:flat? #t)
+       Intersection?)
+   (intern-double-ref!
+    intersection-table
+    elems
+    prop
+    #:construct (make-Intersection ts prop elems))])
 
 (define intersection-table (make-weak-hash))
 
-(define-match-expander Intersection:*
-  (syntax-rules () [(_ ts) (Intersection: ts _)]))
-
 (define (make-Intersection* ts)
   (apply -unsafe-intersect ts))
 
 ;;  constructor for intersections
 ;; in general, intersections should be built
 ;; using the 'intersect' operator, which worries
 ;; about actual subtyping, etc...
-(define (-unsafe-intersect . ts)
-  (let loop ([elems (set)]
-             [ts ts])
-    (match ts
-      [(list)
-       (let ([ts (set->list elems)])
-         (cond
-           [(null? ts) Univ]
-           ;; size = 1 ?
-           [(null? (cdr ts)) (car ts)]
-           ;; size > 1, build an intersection
-           [else (make-Intersection ts elems)]))]
-      [(cons t ts)
-       (match t
-         [(Univ:) (loop elems ts)]
-         [(Intersection: ts* _) (loop elems (append ts* ts))]
-         [_ #:when (for/or ([elem (in-immutable-set elems)]) (not (overlap? elem t)))
-            -Bottom]
-         [_ (loop (set-add elems t) ts)])])))
-
+(define -unsafe-intersect
+  (case-lambda
+    [() Univ]
+    [(t) t]
+    [args
+     (let loop ([ts '()]
+                [elems (hash)]
+                [prop -tt]
+                [args args])
+       (match args
+         [(list)
+          (match ts
+            [(list) (refine/prop Univ prop)]
+            [(list t) (refine/prop (car ts) prop)]
+            [_ (let ([t (make-Intersection ts -tt elems)])
+                 (refine/prop t prop))])]
+         [(cons arg args)
+          (match arg
+            [(Univ:) (loop ts elems prop args)]
+            [(Intersection: ts* (TrueProp:) _) (loop ts elems prop (append ts* args))]
+            [(Intersection: ts* prop* _)
+             (loop ts
+                   elems
+                   (-and prop* prop)
+                   (append ts* args))]
+            [_ #:when (for/or ([elem (in-list args)])
+                        (not (overlap? elem arg)))
+               -Bottom]
+            [t (let ([count (hash-count elems)]
+                     [elems (hash-set elems t #t)])
+                 (if (eqv? count (hash-count elems))
+                     (loop ts elems prop args)
+                     (loop (cons t ts) elems prop args)))])]))]))
+
+(define/provide (Intersection-w/o-prop t)
+  (match t
+    [(Intersection: _ (TrueProp:) _) t]
+    [(Intersection: (list t) prop _) t]
+    [(Intersection: ts prop tset) (make-Intersection ts -tt tset)]))
+
+;; constructs {v ∈ t | prop(v)}
+(define/provide refine/prop
+  (case-lambda
+    [(t prop)
```

Comment:
or if `prop` is already in the "nameless-local" representation form and we just want to put it back together without extra instantiating/abstracting